clippy_utils/src/lib.rs

```diff
@@ -2226,6 +2228,44 @@ pub fn is_expr_used_or_unified(tcx: TyCtxt<'_>, expr: &Expr<'_>) -> bool {
     )
 }
 
+// Checks if the given expression is a call to `align_of`
+// or casting to a type with the same alignment
+pub fn is_expr_const_aligned(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
+    match expr.kind {
+        ExprKind::Call(fun, _) => is_align_of_call(cx, fun, expr),
+        ExprKind::Lit(lit) => is_align_lit(cx, lit, expr),
+        _ => false,
+    }
+}
+
+fn is_align_of_call(cx: &LateContext<'_>, fun: &Expr<'_>, expr: &Expr<'_>) -> bool {
+    let ExprKind::Path(QPath::Resolved(None, path)) = fun.kind else {
+        return false;
+    };
+    let Some(args) = path.segments.last().and_then(|seg| seg.args) else {
+        return false;
+    };
+    let [hir::GenericArg::Type(ty)] = args.args else {
+        return false;
+    };
+    path_def_id(cx, fun).is_some_and(|id| {
+        let typeck = cx.typeck_results();
+        match_any_def_paths(cx, id, &[&paths::CORE_ALIGN_OF, &paths::STD_ALIGN_OF]).is_some()
```

Comment:
The HIR tree always uses the real path of an item, so `std::mem::align_of` will always appear as `core::mem::align_of`.